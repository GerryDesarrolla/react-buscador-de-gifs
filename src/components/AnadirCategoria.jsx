import {useState} from "react";
import PropTypes from "prop-types";

export const AnadirCategoria = ({onNuevaCategoria}) => {
	const [valorInput, setValorInput] = useState('');
	
	const cambiarValorInput = ({target}) => {
		setValorInput(target.value);
	}
	
	const enviarFormulario = (evento) => {
		evento.preventDefault();
		
		const valorLimpio = valorInput.trim();
		
		if(valorLimpio.length <= 1) return;
		
		onNuevaCategoria(valorLimpio);
		
		setValorInput('');
	}
	
	return (
		<form onSubmit={enviarFormulario} aria-label="form">
			<fieldset>
				<input
					autoComplete="off"
					required
					type="text"
					id="buscar"
					value={valorInput}
					onChange={cambiarValorInput}
				/>
				<label
					htmlFor="buscar"
				>Buscar Gifs</label>
			</fieldset>
		</form>
	)
}

AnadirCategoria.protoType = {
	onNuevaCategoria: PropTypes.func.isRequired
}
