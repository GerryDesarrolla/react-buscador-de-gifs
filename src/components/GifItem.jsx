export const GifItem = ({title, image, url}) => {
	return (
		<li className="c-grid__item">
			<a className="c-grid__enlace" href={url} title={title} target="_blank">
				<div className="c-grid__imagen">
					<img className="c-grid__img" src={image} alt={title}/>
				</div>
				<p className="c-grid__titulo">
					{title.length > 0 ? title : 'No tiene t√≠tulo'}
				</p>
			</a>
		</li>
	)
}
