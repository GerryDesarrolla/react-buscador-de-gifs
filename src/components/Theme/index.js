export * from './Loader';
export * from './Theme';
