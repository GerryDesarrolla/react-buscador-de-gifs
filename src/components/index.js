export * from './AnadirCategoria';
export * from './GifGrid';
export * from './GifItem';
