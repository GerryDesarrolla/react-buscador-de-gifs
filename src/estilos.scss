$font-family: Helvetica, Arial, sans-serif;

:root{
	--azul: #191C25;
	--blanco: #ffffff;
}

:root.light{
	--azul: #ffffff;
	--blanco: #191C25;
}

$azul: var(--azul);
$blanco: var(--blanco);

@keyframes moverOnda {
	0% {
		transform: translateX(0) translateZ(0) scaleY(1)
	}
	50% {
		transform: translateX(-25%) translateZ(0) scaleY(0.55)
	}
	100% {
		transform: translateX(-50%) translateZ(0) scaleY(1)
	}
}

*{
	box-sizing: border-box;
}

html, body{
	background: $azul;
	color: $blanco;
	font: 400 1.3rem/2.6rem $font-family;
	margin: 0;
}

h1{
	margin-top: 0;
}

input[type="text"]{
	background: transparent;
	border: none;
	border-bottom: 2px solid $blanco;
	color: $blanco;
	outline: none;
	padding: 10px 0;
	width: 100%;
	
	&:focus, &:valid{
		& + label{
			transform: translateY(-25px) translateX(-35px) scale(0.5);
		}
	}
	
	& ~ label{
		left: 0;
		position: absolute;
		top: 0;
		transition: all 0.5s ease;
	}
}

fieldset{
	border: none;
	margin: 0;
	padding: 0;
	position: relative;
}

.o-container{
	margin: 0 auto;
	max-width: 1240px;
	padding: 100px 20px;
	width: 100%;
}

.o-mensaje{
	background: $blanco;
	border-radius: 30px;
	bottom: 30px;
	color: $azul;
	font: 400 0.9rem/1.5rem $font-family;
	left: 50%;
	margin: 0;
	max-width: 95%;
	opacity: 0;
	padding: 0 30px;
	position: fixed;
	text-align: center;
	transform: translate(-50%, 30px);
	transition: all 0.5s ease;
	visibility: hidden;
	width: 100%;
	
	@media screen and (min-width: 960px) {
		width: auto;
	}
	
	&.is-active{
		opacity: 1;
		transform: translate(-50%, 0);
		visibility: visible;
	}
}

.c-grid{
	display: grid;
	grid-gap: 30px;
	grid-template-columns: repeat(auto-fit, minmax(265px, 1fr));
	list-style: none;
	margin: 0;
	padding: 0;
	width: 100%;
	
	&:hover &{
		&__item{
			opacity: 0.5;
		}
	}
	
	&__enlace{
		text-decoration: none;
	}
	
	&__item{
		background: $blanco;
		border-radius: 6px;
		overflow: hidden;
		transition: all 0.5s ease;
		
		&:hover{
			box-shadow: rgb(38, 57, 77) 0 20px 30px -10px;
			opacity: 1 !important;
			transform: translateY(-5px);
			
			.c-grid{
				&__img{
					transform: scale(1.1);
				}
			}
		}
	}
	
	&__imagen{
		height: 0;
		overflow: hidden;
		padding-bottom: 50%;
		position: relative;
		width: 100%;
	}
	
	&__img{
		height: 100%;
		left: 0;
		object-fit: cover;
		position: absolute;
		top: 0;
		transition: all 0.5s ease;
		width: 100%;
	}
	
	&__titulo{
		-webkit-box-orient: vertical;
		-webkit-line-clamp: 2;
		color: $azul;
		display: -webkit-box;
		font: 400 1rem/1.3rem $font-family;
		margin: 0;
		min-height: 60px;
		overflow: hidden;
		padding: 5px 10px;
		text-overflow: ellipsis;
	}
}
