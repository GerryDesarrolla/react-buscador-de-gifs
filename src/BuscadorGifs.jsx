import {useState} from "react";
import {AnadirCategoria, GiftGrid} from "./components";
import {Theme} from "./components/Theme";

export const BuscadorGifs = () => {
	const [categorias, setCategorias] = useState(['Frontend']);
	const [categoriaRepetida, setCategoriaRepetida] = useState('');
	const [mensajeActivo, setMensajeActivo] = useState(false);
	
	const onAnadirCategoria = (nuevaCategoria) => {
		if(categorias.includes(nuevaCategoria)){
			setCategoriaRepetida(nuevaCategoria);
			setMensajeActivo(true);
			setTimeout(() => setMensajeActivo(false), 3000);
			return;
		}
		setCategorias([nuevaCategoria, ...categorias])
	}
	
	return(
		<>
			<Theme/>
			<div className="o-container">
				<h1>Buscador de Gifs</h1>
				
				<AnadirCategoria onNuevaCategoria={onAnadirCategoria}/>
				
				{categorias.map(categoria => <GiftGrid key={categoria} categoria={categoria}/>)}
				
				<p className={mensajeActivo ? 'o-mensaje is-active' : 'o-mensaje'}>Ya has buscado "{categoriaRepetida}"</p>
			</div>
		</>
	)
}
