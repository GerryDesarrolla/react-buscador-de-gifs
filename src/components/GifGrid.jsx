import {GifItem} from "./GifItem";
import {useFetchGifs} from "../hooks/useFetchGifs";
import {Loader} from "./Theme";
import PropTypes from "prop-types";

export const GiftGrid = ({categoria}) => {
	const {imagenes, isLoading} = useFetchGifs(categoria)
	
	return (
		<>
			{isLoading && <Loader/>}
			
			<h3>{categoria}</h3>
			
			{
				imagenes.length === 0
					? <p>No se encontraron resultado para <i>"{categoria}"</i></p>
					: <ul className="c-grid">
						{imagenes.map((imagen) => (
							<GifItem key={imagen.id} {...imagen}/>
						))}
					</ul>
			}
		</>
	)
}

GiftGrid.propTypes = {
	categoria: PropTypes.string.isRequired
}
