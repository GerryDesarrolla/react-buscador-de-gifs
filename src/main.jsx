import React from 'react';
import ReactDOM from 'react-dom/client';
import {BuscadorGifs} from "./BuscadorGifs";
import "normalize.css/normalize.css";
import "./estilos.scss";

ReactDOM.createRoot(document.getElementById('root')).render(
	<React.StrictMode>
		<BuscadorGifs/>
	</React.StrictMode>
)
