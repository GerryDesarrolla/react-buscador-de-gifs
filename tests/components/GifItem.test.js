import {render, screen} from "@testing-library/react";
import {GifItem} from "../../src/components";

describe('Pruebas en <GifItem/>', () => {
	const imagen = 'https://gerrydesarrolla.com/profile.jpg';
	const titulo = 'Gerry';
	const url = 'https://gerrydesarrolla.com';
	
	test('Debe hacer match con el snapshot', () => {
		const {container} = render(<GifItem image={imagen} title={titulo} url={url}/>);
		
		expect(container).toMatchSnapshot();
	})
	
	test('Debe mostrar la imagen con el URL y el ALT indicado', () => {
		render(<GifItem image={imagen} title={titulo} url={url}/>);
		
		const {src, alt} = screen.getByRole('img');
		
		expect(src).toBe(imagen);
		expect(alt).toBe(titulo);
	})
	
	test('Debe mostrar el titulo en el componente', () => {
		render(<GifItem image={imagen} title={titulo} url={url}/>);
		
		expect(screen.getByText).toBeTruthy();
	})
})
