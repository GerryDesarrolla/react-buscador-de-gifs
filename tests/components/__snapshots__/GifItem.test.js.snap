// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem/> Debe hacer match con el snapshot 1`] = `
<div>
  <li
    class="c-grid__item"
  >
    <a
      class="c-grid__enlace"
      href="https://gerrydesarrolla.com"
      target="_blank"
      title="Gerry"
    >
      <div
        class="c-grid__imagen"
      >
        <img
          alt="Gerry"
          class="c-grid__img"
          src="https://gerrydesarrolla.com/profile.jpg"
        />
      </div>
      <p
        class="c-grid__titulo"
      >
        Gerry
      </p>
    </a>
  </li>
</div>
`;
